test/app-luatest/msgpack_test.lua

```diff
@@ -181,3 +191,227 @@ g.test_encode_decode_uuid = function()
         t.assert_equals(src, res)
     end
 end
+
+g.test_object_encode_decode = function()
+    local o = nil
+    t.assert_equals(msgpack.object(o):decode(), o)
+    local o = box.NULL
+    t.assert_equals(msgpack.object(o):decode(), o)
+    local o = 123
+    t.assert_equals(msgpack.object(o):decode(), o)
+    local o = 'foobar'
+    t.assert_equals(msgpack.object(o):decode(), o)
+    local o = {foo = 123, bar = 456}
+    t.assert_equals(msgpack.object(o):decode(), o)
+    local o = {foo = {1, 2, 3}, bar = {'a', 'b', 'c'}}
+    t.assert_equals(msgpack.object(o):decode(), o)
+    local mp = msgpack.object({bar = 123})
+    t.assert_equals(
+        msgpack.object({mp, {foo = mp}}):decode(),
+        {mp:decode(), {foo = mp:decode()}})
+    t.assert_equals(
+        msgpack.decode(msgpack.encode({mp, {foo = mp}})),
+        {mp:decode(), {foo = mp:decode()}})
+    t.assert_equals(msgpack.object(box.tuple.new(1, 2, 3)):decode(), {1, 2, 3})
+    t.assert_equals(
+        msgpack.object({
+            foo = box.tuple.new(123),
+            bar = box.tuple.new(456),
+        }):decode(), {foo = {123}, bar = {456}})
```

Comment:
The test could look a bit cleaner if it would have form
```Lua
local objs = {
    nil, box.NULL, 123, 'foobar', ...
}
for _, o in pairs(objs) do
    t.assert_equals(msgpack.object(o):decode(), o)
end
```
At least for the exact repetitive parts. Up to you.

Reply:
I intentionally don't use loops in tests so that each assertion can be identified by a line number. Other test cases don't use loops, either (e.g. see `app-luatest/msgpack_test.lua:test_encode_decode_decimals`)

Reply:
Luatest would show mismatching values anyway so finding the failing pair wouldn't be hard, and adding new tests would be easy. But ok.